services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/doi/DoiManagerImpl.java

```diff
@@ -94,45 +94,87 @@ public Doi createOrUpdateDoi(final Long userId, final Doi dto) throws Recoverabl
 		AuthorizationManagerUtil.checkAuthorizationAndThrowException(
 				authorizationManager.canAccess(currentUser, dto.getObjectId(), dto.getObjectType(), ACCESS_TYPE.UPDATE));
 
-		// Check if the DOI exists, and if so, check to make sure eTags match.
+		/*
+		 * If either the association creation fails or the metadata registration fails, we rollback the entire transaction,
+		 *  and either
+		 *  - throw an exception indicates the DOI has already been created (ConflictingUpdateException on eTag mismatch)
+		 *  - throw an exception that suggests retrying the call (RecoverableMessageException)
+		 *
+		 * We do this to avoid a scenario where we register a DOI association, but the metadata is never registered, and
+		 * to avoid certain race conditions, such as
+		 * - the creation of a duplicate association in between retrieval and creation
+		 * - the creation of an association where the external metadata registration fails, and then another association
+		 * 		is created.
+		 *
+		 * In both cases, retrying the original call will succeed if the other caller fails, or fail on a
+		 *  ConflictingUpdateException that indicates that another call successfully minted the DOI.
+		 * Ultimately, we avoid telling the caller that a DOI has been minted when it hasn't.
+		 */
+		DoiAssociation association = createOrUpdateAssociation(dto);
+		dto.setDoiUri(generateDoiUri(dto.getObjectId(), dto.getObjectType(), dto.getObjectVersion()));
+		dto.setDoiUrl(generateLocationRequestUrl(dto.getObjectId(), dto.getObjectType(), dto.getObjectVersion()));
+		DataciteMetadata metadata = createOrUpdateDataciteMetadata(dto);
+		return mergeMetadataAndAssociation(metadata, association);
+	}
+
+	DoiAssociation createOrUpdateAssociation(DoiAssociation dto) throws RecoverableMessageException {
+		DoiAssociation association;
 		try {
+			// Check if the DOI exists by checking to make sure eTags match
+			// (We will get a NotFoundException if it doesn't exist).
 			if (!doiAssociationDao.getEtagForUpdate(dto.getObjectId(), dto.getObjectType(), dto.getObjectVersion()).equals(dto.getEtag())) {
 				throw new ConflictingUpdateException("eTags do not match.");
 			}
-			doiAssociationDao.updateDoiAssociation(dto);
-		} catch (NotFoundException e1) {
+			association = doiAssociationDao.updateDoiAssociation(dto);
+		} catch (NotFoundException e1) { // The DOI does not already exist (exception was thrown by getEtag)
 			try {
-				doiAssociationDao.createDoiAssociation(dto); //retry on fail
-			} catch (Exception e2) {
+				association = doiAssociationDao.createDoiAssociation(dto); // Create
+			} catch (DatastoreException e2) {
+				/*
+				 * If the create method fails after encountering the NotFoundException, it is possible that a separate
```

Comment:
This exception indicates there was a race condition where two callers attempted to create same DOI at the same time.  The loser of this race will see this exception.  However, since the winner might also fail before completion, we send this caller back to the beginning to retry.